### ISSUE TITLE
Create a design document for adding the functionality of player movement throughout the world

### ISSUE THREAD

**Issue by philipchenaux** [uchicago-cs/chiventure#468]
Another npc independent feature that we will focus upon is the ability of npcs to freely navigate the world on their own. This will be dependent on the npc module created by the npc team. This will requrie some discussion with npc team as to who may be responsible for this task, but the implementation and design will be covered on a wiki page.

**Comment by philipchenaux** [comment]
Following a team meeting with the npc team, we were able to define and desperate tasks accordingly for each team. The full description of the meeting is covered in #463 .

Therefore we will commence the design phase of npc player movement and work on producing a design document to be posted on the wiki by the end of this sprint.

**Comment by philipchenaux** [comment]
Following a brief meeting with the rpg open-world team to discuss the implementation of npc movement, I was able to clarify a few basic elements of how they are planning to setup and structure the world. Below are the key takeaways from the meeting that relate to this issue:

- Levels - which npcs would be there in which level / room
- Need to think about what kind of NPCs (Hostile or non hostile one) will be generated in a room. 

A complete rundown on what was discussed in the meeting can be found here:
https://docs.google.com/document/d/1KBx7OtPcT0U4NnX1G47J30kuNH_cYu6OUyHTHTsg5mg/edit

Unfortunately given that this meeting also addressed other dependencies between rpg teams such as rpg-playerclass, I was not able to fully flesh out the ideas. I hope to further the clarification with the rpg-openworld team next sprint.

**Comment by philipchenaux** [comment]
I've started working on a wiki that outlines a rough design for two key npc features that will be part of the open-world module. First is the ability of npcs to spawn in different room types, and the other feature is to enable npc movement between the rooms. To construct this, I had to reference both the npcs wiki:
https://github.com/uchicago-cs/chiventure/wiki/NPC-Design-and-Planning
as well as the open-worlds sample_room module:
https://github.com/uchicago-cs/chiventure/blob/openworld/sample-generation/include/openworld/sample_rooms.h

The key details from my design document are:

- To implement the npc spawn feature: Each room is associated with a room_tag which specifies what type of room it is, similarly each npc has a enum that specifies their class npc_class. The first step would be to associate an npc_class with a room_tag (but note that an npc_class can be associate with several different room_tags). 
 - A potential other feature we can implement would be to integrate an npc movement element in which it can movement through adjacent rooms similar to how the player can. It would require first identifying the room_tags of each of the adjacent rooms and then determining whether the room_tags of the adjacent rooms are associated to the player_class of the npc. And if so then the npc has the ability to move (but this raises the question: when should the npc be moving rooms? Maybe at regular time intervals? Or after a certain event? Or randomly?)

**Comment by philipchenaux** [comment]
Following this design document, I am still awaiting a response from the other teams, but I will begin a coding an implementation of these features in the following sprint. I have a rough design of the structs and high-level functions of how I wish to implement this. One issue I had were the holes that existed in the npc and open-world module that I had to fill with my own ideas. I will discuss with both teams to see if my ideas fit in accordingly before starting the implementation, but I will close this issue for now and reopen another issue that addresses the two features separately.

**Comment by firatciftci** [comment]
Issue Score: ✔️++

Comments:
This issue looks great overall. I particularly like the extensive update comments you have listed. Keep up the good work!
